python/lsst/ip/isr/ampOffset.py

```diff
@@ -18,88 +18,406 @@
+            bgImage = bg.getImageF(self.background.config.algorithm, self.background.config.undersampleStyle)
+            maskedImage -= bgImage
+
+        # Detect sources and update cloned exposure mask planes in-place.
+        if self.config.doDetection:
+            schema = SourceTable.makeMinimalSchema()
+            table = SourceTable.make(schema)
+            # Detection sigma, used for smoothing and to grow detections, is
+            # normally measured from the PSF of the exposure. As the PSF hasn't
+            # been measured at this stage of processing, sigma is instead
+            # set to an approximate value here (which should be sufficient).
+            _ = self.detection.run(table=table, exposure=exp, sigma=2)
+
+        # Safety check: do any pixels remain for amp offset estimation?
+        if (exp.mask.array & bitMask).all():
+            self.log.warning("All pixels masked: cannot calculate any amp offset corrections.")
+        else:
+            # Set up amp offset inputs.
+            im = exp.image
+            im.array[(exp.mask.array & bitMask) > 0] = np.nan
+            amps = exp.getDetector().getAmplifiers()
+
+            min_dimension = np.min([np.min(amp.getBBox().getDimensions()) for amp in amps])
+            if self.config.ampEdgeWindow > min_dimension:
+                raise RuntimeError(
+                    f"The sliding window ({self.config.ampEdgeWindow} pixels) is larger than the smaller "
+                    f"dimension in amplifiers ({min_dimension} pixels). This can lead to downstream issues "
+                    "after convolution in the `getInterfaceAmpOffset()` method. Please adjust "
+                    f"`ampEdgeWindow` in the config to a value equal to or less than {min_dimension} pixels "
+                    "and rerun."
+                )
+
+            # Determine amplifier interface geometry.
+            ampAreas = {amp.getBBox().getArea() for amp in amps}
+            if len(ampAreas) > 1:
+                raise NotImplementedError(
+                    "Amp offset correction is not yet implemented for detectors with differing amp sizes."
+                )
+
+            # Obtain association and offset matrices for the subsequent
+            # operation.
+            A, interfaces = self.getAmpAssociations(amps)
+            B = self.getAmpOffsets(im, amps, A, interfaces)
+
+            # If least-squares minimization fails, convert NaNs to zeroes,
+            # ensuring that no values are erroneously added/subtracted
+            pedestals = np.nan_to_num(np.linalg.lstsq(A, B, rcond=None)[0])
+            metadata = exposure.getMetadata()
+            for ii, (amp, pedestal) in enumerate(zip(amps, pedestals)):
+                ampIm = exposure.image[amp.getBBox()].array
+                ampIm -= pedestal
+                metadata.set(
+                    f"PEDESTAL{ii + 1}", float(pedestal), f"Pedestal level subtracted from amp {ii + 1}"
```

Comment:
This  +1 will be incorrect for other cameras. In HSC, it was true that the zeroth amp had an ID of 1, the first amp had an ID of 2, and so on. 

For this generalized task, we'll need to figure out a way to pass the actual amp ID into this message, and propagate it through.

Reply:
This should just be `amp.getId()` from the amplifier object.  `.getName()` may also be useful if this is for humans.

Reply:
amp.getName() worked!